How many late days are you using for this assignment? 0

In [55]:
# Imports and pip installations (if needed)
import sklearn as sk
import pandas as pd


# Part 1: Load the dataset

In [71]:
# Load the given datasets
categorical_data = pd.read_csv("chronic_kidney_disease_categorical.csv")
numerical_data = pd.read_csv("chronic_kidney_disease_numerical.csv")

# Print the categorical data
print("Categorical Data:")
print(categorical_data)

# Print the numerical data
print("\nNumerical Data:")
print(numerical_data)

Categorical Data:
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  go

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: These datasets are a collection of medical data which can be used to predict if patients have Chronic Kidney Disease. The features are the input variables which are used to predict whether patients have Chronic Kidney Disease which can include categorical and numerical. The labels are the target variables which can indicate if the patient had Chronic Kidney Disease or not; these are ckd or notckd, respectively. These features are both categorical as well as numerical. Examples of categorical features are rbc (Red Blood Cells), pc (Pus Cells), and pcc (Pus Cell Clumps). Examples of numerical features are age, bp (Blood Pressure), and bgr (Blood Glucose Random). 

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [72]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows


# Drop these duplicate rows


# Repeat the same for categorical dataset. Print the duplicate rows and drop them

# Check for duplicate rows in the numerical dataset
numerical_duplicates = numerical_data.duplicated()
total_numerical_duplicates = numerical_duplicates.sum()
print("Total duplicate rows in numerical dataset:", total_numerical_duplicates)

# Print duplicate rows in the numerical dataset
if total_numerical_duplicates > 0:
    print("Duplicate rows in numerical dataset:")
    print(numerical_data[numerical_duplicates])

# Drop the duplicate rows in the numerical dataset
numerical_data = numerical_data.drop_duplicates()

# Check for duplicate rows in the categorical dataset
categorical_duplicates = categorical_data.duplicated()
total_categorical_duplicates = categorical_duplicates.sum()
print("Total duplicate rows in categorical dataset:", total_categorical_duplicates)

# Print duplicate rows in the categorical dataset
if total_categorical_duplicates > 0:
    print("Duplicate rows in categorical dataset:")
    print(categorical_data[categorical_duplicates])

# Drop the duplicate rows in the categorical dataset
categorical_data = categorical_data.drop_duplicates()


Total duplicate rows in numerical dataset: 30
Duplicate rows in numerical dataset:
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
107     872034  70.0   90.0  184.0   98.6  3.3  138.0  3.9   5.8   NaN   
149     721270  58.0   70.0  102.0   48.0  1.2  139.0  4.3  15.0  40.0   
171     240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
173     441170  79.0   80.0  111.0   44.0  1.2  146.0  3.6  16.3  40.0   
185     261137  35.0   60.0  105.0   39.0  0.5  135.0  3.9  14.7  43.0   
194     769512  38.0   60.0   91.0   36.0  0.7  135.0  3.7  14.0  46.0   
213     354223  64.0   70.0   97.0   27.0  0.7  145.0  4.8  13.8  49.0   
244     302906  59.0  100.0    NaN   96.0  6.4    NaN  NaN   6.6   NaN   
245     431871  57.0   60.0  132.0   18.0  1.1  150.0  4.7  15.4  42.0   
248     687021  52.0   80.0   99.0   25.0  0.8  135.0  3.7  15.0  52.0   
249     843449  30.0   80.0  131.0   38.0  1.0  147.0  3.8  14.1  45.0   
265     628271  65.0   90.0  

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: As we can see from the output, all the rows with unique ids are present in both datasets. This is due to the 
natrue of the merge we performed being an outer join and all the unique ids in both datasets are being used. This
aligns with the resulting shape of our table being 400x25.

In [73]:
# Merge the two given numerical and categorical datasets based on their unique_ID.

#Print the combined dataset

# Merge the datasets based on unique_ID
combined_data = pd.merge(numerical_data, categorical_data, on='unique_id', how='outer')

# Print the combined dataset
print("Combined Dataset:")
combined_data



Combined Dataset:


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,...,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
395,298297,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
396,948317,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
397,488947,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,abnormal,notpresent,notpresent,no,no,no,good,no,no,ckd
398,663387,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,normal,notpresent,notpresent,no,no,no,poor,no,yes,ckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [74]:
# Calculate the percentage of rows that contain atleast one missing value

# Print %

# Drop these rows from the dataset

# Print the Dataset


# Calculate the percentage of rows with at least one missing value
missing_rows_percentage = (combined_data.isnull().any(axis=1).sum() / combined_data.shape[0]) * 100

# Print the percentage
print(f"Percentage of rows that contain at least one missing value: {missing_rows_percentage:.2f}%")

# Drop the rows with any missing values
combined_data_cleaned = combined_data.dropna()

# Print the final dataset after dropping missing values
print("\nDataset after dropping rows with missing values:")
combined_data_cleaned

Percentage of rows that contain at least one missing value: 61.75%

Dataset after dropping rows with missing values:


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
7,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
13,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,...,abnormal,notpresent,notpresent,yes,yes,yes,good,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
379,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
381,462892,36.0,80.0,85.0,16.0,1.1,142.0,4.1,15.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
386,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,...,abnormal,notpresent,present,yes,yes,no,good,yes,no,ckd
387,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.4: Sort the dataset according to the Labels

In [75]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting

# Print the dataset

# Sort the dataset according to the values in 'Target' column and reset indices
sorted_data = combined_data_cleaned.sort_values(by='Target').reset_index(drop=True)

# Print the sorted dataset
print("Sorted Dataset:")
sorted_data

Sorted Dataset:


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,abnormal,present,present,no,no,no,good,no,yes,ckd
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,...,abnormal,present,notpresent,yes,yes,yes,poor,no,no,ckd
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,...,abnormal,present,notpresent,yes,yes,no,poor,no,yes,ckd
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,61.0,70.0,120.0,29.0,0.7,137.0,3.5,17.4,52.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
149,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,614376,57.0,80.0,119.0,17.0,1.2,135.0,4.7,15.4,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
151,853426,41.0,70.0,125.0,38.0,0.6,140.0,5.0,16.8,41.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [76]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

encoded_data = pd.get_dummies(sorted_data, columns=categorical_columns)

# Print the encoded dataset
print("Encoded Dataset:")
encoded_data
# Print the dataset

Encoded Dataset:


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,True,False,True,False,True,False,False,True,True,False
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,...,False,True,False,True,True,False,True,False,True,False
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,...,True,False,False,True,True,False,False,True,True,False
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,False,True,False,True,False,True,False,True,True,False
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,...,True,False,True,False,True,False,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,61.0,70.0,120.0,29.0,0.7,137.0,3.5,17.4,52.0,...,True,False,True,False,True,False,True,False,False,True
149,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,...,True,False,True,False,True,False,True,False,False,True
150,614376,57.0,80.0,119.0,17.0,1.2,135.0,4.7,15.4,42.0,...,True,False,True,False,True,False,True,False,False,True
151,853426,41.0,70.0,125.0,38.0,0.6,140.0,5.0,16.8,41.0,...,True,False,True,False,True,False,True,False,False,True


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: This solution of using a single column with a binary variable is also a valid way to represent the value. This would be effective for just 2 possible values. With multiple possible values, then this would prove to have complications with the model not being able to separate different values which can only be represented with a 0 or 1. Thus, for categorical variables which can be more than 2 values, using dummy variables would be more effective for the machine to learn. 

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [77]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

# Remove outliers from the numerical columns
for column in numerical_columns:
    mean = encoded_data[column].mean()
    std_dev = encoded_data[column].std()
    # Define the upper and lower bounds for outliers
    lower_bound = mean - 3 * std_dev
    upper_bound = mean + 3 * std_dev
    # Filter the dataset to remove outliers
    encoded_data = encoded_data[(encoded_data[column] >= lower_bound) & (encoded_data[column] <= upper_bound)]

# Print the dataset after removing outliers
print("Dataset after removing outliers:")
encoded_data
# Print the dataset

Dataset after removing outliers:


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,True,False,True,False,True,False,False,True,True,False
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,...,False,True,False,True,True,False,True,False,True,False
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,False,True,False,True,False,True,False,True,True,False
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,...,True,False,True,False,True,False,True,False,True,False
6,546225,46.0,60.0,163.0,92.0,3.3,141.0,4.0,9.8,28.0,...,True,False,True,False,True,False,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,61.0,70.0,120.0,29.0,0.7,137.0,3.5,17.4,52.0,...,True,False,True,False,True,False,True,False,False,True
149,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,...,True,False,True,False,True,False,True,False,False,True
150,614376,57.0,80.0,119.0,17.0,1.2,135.0,4.7,15.4,42.0,...,True,False,True,False,True,False,True,False,False,True
151,853426,41.0,70.0,125.0,38.0,0.6,140.0,5.0,16.8,41.0,...,True,False,True,False,True,False,True,False,False,True


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [78]:
# Normalize the all Numerical Attributes in the dataset.

# Print the dataset

# Normalize the numerical columns using Min-Max scaling
for column in numerical_columns:
    encoded_data[column] = (encoded_data[column] - encoded_data[column].min()) / (encoded_data[column].max() - encoded_data[column].min())

# Print the dataset after normalization
print("Dataset after normalizing numerical attributes:")
encoded_data

Dataset after normalizing numerical attributes:


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.206897,0.059406,0.000000,...,True,False,True,False,True,False,False,True,True,False
1,137148,0.905405,1.00,0.965665,0.522876,0.641975,0.666667,0.000000,0.148515,0.225806,...,False,True,False,True,True,False,True,False,True,False
3,343710,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,0.032258,...,False,True,False,True,False,True,False,True,True,False
4,532520,0.729730,0.75,0.150215,0.281046,0.234568,0.533333,0.793103,0.336634,0.322581,...,True,False,True,False,True,False,True,False,True,False
6,546225,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.379310,0.207921,0.161290,...,True,False,True,False,True,False,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,0.743243,0.25,0.214592,0.124183,0.037037,0.566667,0.206897,0.960396,0.935484,...,True,False,True,False,True,False,True,False,False,True
149,125587,0.418919,0.00,0.167382,0.241830,0.086420,0.700000,0.689655,0.722772,0.806452,...,True,False,True,False,True,False,True,False,False,True
150,614376,0.689189,0.50,0.210300,0.045752,0.098765,0.500000,0.620690,0.762376,0.612903,...,True,False,True,False,True,False,True,False,False,True
151,853426,0.472973,0.25,0.236052,0.183007,0.024691,0.666667,0.724138,0.900990,0.580645,...,True,False,True,False,True,False,True,False,False,True


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The unique id column in the dataset is not relevant for modelling amd making predictions for Chronic Kidney Disease. This merely serves as an identifier and is not required as we go about data manipulation. Including this may or may not confuse the model and it is irrelevant for our goals. 

In [79]:
#Remove that column

# Print the dataset

# Remove the identified columns
cleaned_data = encoded_data.drop(columns="unique_id")

# Print the dataset after removing inappropriate columns
print("Dataset after removing inappropriate columns:")
cleaned_data

Dataset after removing inappropriate columns:


,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.206897,0.059406,0.000000,0.653226,...,True,False,True,False,True,False,False,True,True,False
1,0.905405,1.00,0.965665,0.522876,0.641975,0.666667,0.000000,0.148515,0.225806,0.217742,...,False,True,False,True,True,False,True,False,True,False
3,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,0.032258,0.395161,...,False,True,False,True,False,True,False,True,True,False
4,0.729730,0.75,0.150215,0.281046,0.234568,0.533333,0.793103,0.336634,0.322581,0.500000,...,True,False,True,False,True,False,True,False,True,False
6,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.379310,0.207921,0.161290,0.830645,...,True,False,True,False,True,False,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,0.743243,0.25,0.214592,0.124183,0.037037,0.566667,0.206897,0.960396,0.935484,0.217742,...,True,False,True,False,True,False,True,False,False,True
149,0.418919,0.00,0.167382,0.241830,0.086420,0.700000,0.689655,0.722772,0.806452,0.217742,...,True,False,True,False,True,False,True,False,False,True
150,0.689189,0.50,0.210300,0.045752,0.098765,0.500000,0.620690,0.762376,0.612903,0.153226,...,True,False,True,False,True,False,True,False,False,True
151,0.472973,0.25,0.236052,0.183007,0.024691,0.666667,0.724138,0.900990,0.580645,0.161290,...,True,False,True,False,True,False,True,False,False,True


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [80]:
# Export the dataframe to a new csv file

# Export the dataframe to a new json file

# Export the cleaned DataFrame to a CSV file
cleaned_data.to_csv("cleaned_dataset.csv", index=False)

# Export the cleaned DataFrame to a JSON file
cleaned_data.to_json("cleaned_dataset.json", orient='records', lines=True)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [81]:
# Output the top 15 rows
top_15_rows = cleaned_data.head(15)
# Output the top 15 rows with reset indices
top_15_rows = cleaned_data.head(15).reset_index(drop=True)
# Print the top 15 rows
top_15_rows

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.206897,0.059406,0.000000,0.653226,...,True,False,True,False,True,False,False,True,True,False
1,0.905405,1.00,0.965665,0.522876,0.641975,0.666667,0.000000,0.148515,0.225806,0.217742,...,False,True,False,True,True,False,True,False,True,False
2,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,0.032258,0.395161,...,False,True,False,True,False,True,False,True,True,False
3,0.729730,0.75,0.150215,0.281046,0.234568,0.533333,0.793103,0.336634,0.322581,0.500000,...,True,False,True,False,True,False,True,False,True,False
4,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.379310,0.207921,0.161290,0.830645,...,True,False,True,False,True,False,True,False,True,False
5,0.675676,0.75,0.253219,0.633987,0.777778,0.366667,0.655172,0.138614,0.193548,0.169355,...,True,False,True,False,True,False,True,False,True,False
6,0.716216,0.50,1.000000,0.163399,0.111111,0.066667,0.206897,0.267327,0.387097,0.532258,...,True,False,False,True,True,False,True,False,True,False
7,0.729730,0.00,0.935622,0.169935,0.160494,0.333333,0.034483,0.019802,0.064516,0.879032,...,True,False,False,True,True,False,False,True,True,False
8,0.000000,0.00,0.103004,0.372549,0.074074,0.500000,0.689655,0.217822,0.225806,1.000000,...,True,False,False,True,True,False,True,False,True,False
9,0.878378,0.00,0.206009,0.751634,0.604938,0.533333,0.689655,0.366337,0.387097,0.879032,...,True,False,False,True,False,True,True,False,True,False


|   age   |    bp   |     bgr     |    bu     |    sc     |    sod    |    pot    |   hemo    |   pcv    |   wbcc   | ... | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|---------|---------|-------------|-----------|-----------|-----------|-----------|-----------|----------|----------|-----|--------|---------|------------|------------|-------|--------|--------|---------|-------------|----------------|
| 0.897059| 1.00    | 0.965665    | 0.606061  | 0.825397  | 0.666667  | 0.000000  | 0.011494  | 0.076923 | 0.247706 | ... | False  | True    | False      | True       | True  | False  | True   | False   | True        | False          |
| 0.691176| 0.50    | 1.000000    | 0.189394  | 0.142857  | 0.066667  | 0.206897  | 0.149425  | 0.269231 | 0.605505 | ... | True   | False   | False      | True       | True  | False  | True   | False   | True        | False          |
| 0.500000| 0.00    | 0.399142    | 0.621212  | 0.460317  | 0.700000  | 0.379310  | 0.080460  | 0.000000 | 0.944954 | ... | True   | False   | True       | False      | True  | False  | True   | False   | True        | False          |
| 0.705882| 0.75    | 0.150215    | 0.325758  | 0.301587  | 0.533333  | 0.793103  | 0.229885  | 0.192308 | 0.568807 | ... | True   | False   | True       | False      | True  | False  | True   | False   | True        | False          |
| 0.647059| 0.75    | 0.253219    | 0.734848  | 1.000000  | 0.366667  | 0.655172  | 0.000000  | 0.038462 | 0.192661 | ... | True   | False   | True       | False      | True  | False  | True   | False   | True        | False          |
| 0.867647| 0.00    | 0.206009    | 0.871212  | 0.777778  | 0.533333  | 0.689655  | 0.264368  | 0.269231 | 1.000000 | ... | True   | False   | False      | True       | False | True   | True   | False   | True        | False          |
| 0.867647| 0.25    | 0.639485    | 0.545455  | 0.507937  | 0.433333  | 0.517241  | 0.149425  | 0.192308 | 0.119266 | ... | False  | True    | True       | False      | True  | False  | True   | False   | True        | False          |
| 0.838235| 0.25    | 0.618026    | 0.651515  | 0.936508  | 0.000000  | 0.344828  | 0.034483  | 0.000000 | 0.660550 | ... | False  | True    | True       | False      | False | True   | False  | True    | True        | False          |
| 0.750000| 1.00    | 0.901288    | 0.189394  | 0.444444  | 0.766667  | 0.206897  | 0.448276  | 0.461538 | 0.504587 | ... | False  | True    | True       | False      | True  | False  | True   | False   | True        | False          |
| 0.897059| 0.50    | 0.785408    | 1.000000  | 0.666667  | 0.600000  | 1.000000  | 0.160920  | 0.192308 | 0.266055 | ... | False  | True    | True       | False      | True  | False  | True   | False   | True        | False          |
| 0.632353| 0.50    | 0.618026    | 0.477273  | 0.555556  | 0.566667  | 0.689655  | 0.206897  | 0.230769 | 0.284404 | ... | True   | False   | True       | False      | False | True   | True   | False   | True        | False          |
| 0.735294| 0.25    | 0.223176    | 0.242424  | 0.206349  | 0.533333  | 0.620690  | 0.402299  | 0.423077 | 0.330275 | ... | True   | False   | True       | False      | True  | False  | True   | False   | True        | False          |
| 0.764706| 0.00    | 0.725322    | 0.363636  | 0.619048  | 0.566667  | 0.862069  | 0.045977  | 0.038462 | 0.293578 | ... | True   | False   | False      | True       | False | True   | True   | False   | True        | False          |
| 0.838235| 0.25    | 0.832618    | 0.583333  | 0.365079  | 0.333333  | 0.379310  | 0.390805  | 0.346154 | 0.486239 | ... | False  | True    | True       | False      | False | True   | True   | False   | True        | False          |
| 0.647059| 0.25    | 0.600858    | 0.121212  | 0.206349  | 0.533333  | 0.310345  | 0.804598  | 0.923077 | 0.752294 | ... | True   | False   | True       | False      | True  | False  | True   | False   | True        | False          |


### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

For bigger datasets, it is reasonable to assume that these LLMs would "hallucinate" during these data conversion tasks. To mitigate these, possible solutions could be to implement automated testing which verifies that the output of these LLMs are in fact accurate by cross referencing with previous valid datasets. This would have a layer of protection so as to ensure that there was no mistakes during the process. Another possible solution could be testing using smaller sample sizes of your dataset and not overwhelm the LLM. 

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [82]:
# Code to rename all the columns in the dataset
# Create a mapping dictionary for renaming the columns
# Create a mapping dictionary for renaming the columns
# Create a mapping dictionary for renaming the columns using full descriptive names
rename_mapping = {
    'age': 'Age',
    'bp': 'Blood Pressure',
    'bgr': 'Blood Glucose Random ',
    'bu': 'Blood Urea',
    'sc': 'Serum Creatinine',
    'sod': 'Sodium',
    'pot': 'Potassium',
    'hemo': 'Hemoglobin',
    'pcv': 'Packed Cell Volume',
    'wbcc': 'White Blood Cell Count',
    'rbcc': 'Red Blood Cell Count',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc_abnormal': 'Red Blood Cell Abnormal',
    'rbc_normal': 'Red Blood Cell Normal ',
    'pc_abnormal': 'Pus Cells Abnormal',
    'pc_normal': 'Pus Cells Normal',
    'pcc_notpresent': 'Pus Cell Clump Not Present',
    'pcc_present': 'Pus Cell Clump Present',
    'ba_notpresent': 'Bacteria Not Present',
    'ba_present': 'Bacteria Present',
    'htn_no': 'Hypertension No',
    'htn_yes': 'Hypertension Yes',
    'dm_no': 'Diabetes Mellitus No',
    'dm_yes': 'Diabetes Mellitus Yes',
    'cad_no': 'Coronary Artery Disease No',
    'cad_yes': 'Coronary Artery Disease Yes',
    'appet_good': 'Appetite Good',
    'appet_poor': 'Appetite Poor',
    'pe_no': 'Pedal Edema No',
    'pe_yes': 'Pedal Edema Yes',
    'ane_no': 'Anemia No',
    'ane_yes': 'Anemia Yes',
    'Target_ckd': 'Chronic Kidney Disease',
    'Target_notckd': 'Not Chronic Kidney Disease'
}

# Rename the columns using the mapping
cleaned_data.rename(columns=rename_mapping, inplace=True)

# Print the DataFrame to confirm changes
cleaned_data

   

,Age,Blood Pressure,Blood Glucose Random,Blood Urea,Serum Creatinine,Sodium,Potassium,Hemoglobin,Packed Cell Volume,White Blood Cell Count,...,Coronary Artery Disease No,Coronary Artery Disease Yes,Appetite Good,Appetite Poor,Pedal Edema No,Pedal Edema Yes,Anemia No,Anemia Yes,Chronic Kidney Disease,Not Chronic Kidney Disease
0,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.206897,0.059406,0.000000,0.653226,...,True,False,True,False,True,False,False,True,True,False
1,0.905405,1.00,0.965665,0.522876,0.641975,0.666667,0.000000,0.148515,0.225806,0.217742,...,False,True,False,True,True,False,True,False,True,False
3,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,0.032258,0.395161,...,False,True,False,True,False,True,False,True,True,False
4,0.729730,0.75,0.150215,0.281046,0.234568,0.533333,0.793103,0.336634,0.322581,0.500000,...,True,False,True,False,True,False,True,False,True,False
6,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.379310,0.207921,0.161290,0.830645,...,True,False,True,False,True,False,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,0.743243,0.25,0.214592,0.124183,0.037037,0.566667,0.206897,0.960396,0.935484,0.217742,...,True,False,True,False,True,False,True,False,False,True
149,0.418919,0.00,0.167382,0.241830,0.086420,0.700000,0.689655,0.722772,0.806452,0.217742,...,True,False,True,False,True,False,True,False,False,True
150,0.689189,0.50,0.210300,0.045752,0.098765,0.500000,0.620690,0.762376,0.612903,0.153226,...,True,False,True,False,True,False,True,False,False,True
151,0.472973,0.25,0.236052,0.183007,0.024691,0.666667,0.724138,0.900990,0.580645,0.161290,...,True,False,True,False,True,False,True,False,False,True


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [83]:
# Converted SQL to Pandas code
# Group by 'Target' column and count occurrences
target_count = cleaned_data.groupby('Chronic Kidney Disease').size().reset_index(name='count')

# Display the resulting DataFrame
target_count

,Chronic Kidney Disease,count
0,False,111
1,True,22


*Describe what the above code does here*
The code above does the following - 
- Groups the rows of the dataframe by the unique values found in Chronic Kidney Disease Column 
- Counts the occurrences using size 
- Then converts the Series to dataframe again and now count becomes a new column
- Puts these in new dataframe called target_count and displays this